File: Backend/Controllers/CalculatorController.cs
using Backend.Models;
using LogicServices;
using Microsoft.AspNetCore.Mvc;
using System.Diagnostics;

namespace Backend.Controllers
{
   public class CalculatorController : Controller
   {
      private DescriptiveStatistics statsSvc;
      private LinearRegression linearRegSvc;

      // ----------------------------------------------------------------------
      //
      public CalculatorController() 
      {
         statsSvc = new DescriptiveStatistics();
         linearRegSvc = new LinearRegression();
      }

      // ----------------------------------------------------------------------
      //
      public IActionResult Index()
      {
         return View("Calculator");
      }

      // ----------------------------------------------------------------------
      //
      [ResponseCache(Duration = 0, Location = ResponseCacheLocation.None, NoStore = true)]
      public IActionResult Error()
      {
         return View(new ErrorViewModel { RequestId = Activity.Current?.Id ?? HttpContext.TraceIdentifier });
      }

      // ----------------------------------------------------------------------
      //
      [HttpPost]
      public CalculatorResponse CalculateStdDev([FromBody] string data)
      {
         List<double> values;
         bool isValid = statsSvc.IsDataValidForSampleStandardDeviation(data, out values);
         if (! isValid)
            return (new CalculatorError (false, "Invalid Input", "Standard Deviation format is one number per line"));
         else
         {
            double sd = statsSvc.StandardDeviation(values);
            return (new CalculatorAnswer<double>(true, "Sample Standard Deviation", sd));
         }
      }

      // ----------------------------------------------------------------------
      //
      [HttpPost]
      public CalculatorResponse CalculatePopulationStdDev([FromBody] string data)
      {
         List<double> values;
         bool isValid = statsSvc.IsDataValidForPopulationStandardDeviation(data, out values);
         if (!isValid)
            return (new CalculatorError(false, "Invalid Input", "Standard Deviation format is one number per line"));
         else
         {
            double psd = statsSvc.PopulationStandardDeviation(values);
            return (new CalculatorAnswer<double>(true, "Population Standard Deviation", psd));
         }
      }

      // ----------------------------------------------------------------------
      //
      [HttpPost]
      public CalculatorResponse CalculateMean([FromBody] string data)
      {
         List<double> values;
         bool isValid = statsSvc.IsDataValidForMean(data, out values);
         if (!isValid)
            return (new CalculatorError(false, "Invalid Input", "Mean format is one number per line"));
         else
         {
            double mean = statsSvc.Mean(values);
            return (new CalculatorAnswer<double>(true, "Mean", mean));
         }
      }

      // ----------------------------------------------------------------------
      //
      [HttpPost]
      public CalculatorResponse CalculateZScore([FromBody] string data)
      {
         if (! statsSvc.IsDataValidForZScore(data, out double value, out double mean, out double sd))
            return (new CalculatorError(false, "Invalid Input", "Z-Score format is \"value, mean, stdDev\" on one line separated by commas"));
         {
            double zScore = statsSvc.ZScore(value, mean, sd);
            return (new CalculatorAnswer<double>(true, "Z-Score", zScore));
         }
      }

      // ----------------------------------------------------------------------
      //
      [HttpPost]
      public CalculatorResponse CalculateLinearRegression([FromBody] string data)
      {
         List<XYPair> values;
         if (!linearRegSvc.IsDataValidForLinearRegression(data, out values))
            return (new CalculatorError(false, "Invalid Input", "Simple Linear Regression format is one x, y value per line"));
         {
            string formula = linearRegSvc.CalculateLinearRegression(values);
            return (new CalculatorAnswer<string>(true, "Single Linear Regression Formula", formula));
         }
      }

      // ----------------------------------------------------------------------
      //
      [HttpPost]
      public CalculatorResponse CalculateYValue([FromBody] string data)
      {
         if (!linearRegSvc.IsDataValidForYValue(data, out double x, out double m, out double b))
            return (new CalculatorError(false, "Invalid Input", "Predict Y format is \"x,m,b\" on one line separated by commas"));
         {
            double yValue = linearRegSvc.CalculateYValue(x, m, b);
            return (new CalculatorAnswer<double>(true, "Single Linear Regression Prediction", yValue));
         }
      }
   }
}

File: calculator/src/main/java/com/calculator/controllers/CalculatorController.java
package com.calculator.controllers;

import java.util.List;

import com.calculator.models.*;
import com.calculatorlogic.*;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class CalculatorController
{
   private DescriptiveStatistics statsSvc;
   private LinearRegression linearRegSvc;

   // ----------------------------------------------------------------------
   //
   public CalculatorController() 
   {
      statsSvc = new DescriptiveStatistics();
      linearRegSvc = new LinearRegression();
   }

   // ----------------------------------------------------------------------
   //
   private String toJSON (Object o)
   {
      ObjectMapper om = new ObjectMapper();
      String json = null;
      try
      {
         json = om.writeValueAsString (o);
      }
      catch (Exception e)
      {
      }

      return (json);
   }
   
   // ----------------------------------------------------------------------
   //
   @PostMapping(path="/CalculateStdDev", consumes = MediaType.APPLICATION_JSON_VALUE,
           produces = MediaType.APPLICATION_JSON_VALUE)
   public String CalculateStdDev(@RequestBody String data)
   {
      //ZScoreInput i = om.readValue(data, ZScoreInput.class);
      List<Double> values = statsSvc.IsDataValidForSampleStandardDeviation(data);
      if (values == null)
      {
         CalculatorError error = new CalculatorError(false, "Invalid Input", "Standard Deviation format is one number per line");
         return (toJSON (error));
      }
      else
      {
         Double sd = statsSvc.StandardDeviation(values);
         CalculatorAnswer answer = new CalculatorAnswer(true, "Sample Standard Deviation", sd);
         return (toJSON (answer));
      }
   }

   // ----------------------------------------------------------------------
   //
   @PostMapping(path="/CalculatePopulationStdDev", consumes = MediaType.APPLICATION_JSON_VALUE,
           produces = MediaType.APPLICATION_JSON_VALUE)
   public String CalculatePopulationStdDev(@RequestBody String data)
   {
      List<Double> values = statsSvc.IsDataValidForPopulationStandardDeviation(data);
      if (values == null) {
         CalculatorError error = new CalculatorError(false, "Invalid Input", "Pop-Standard Deviation format is one number per line");
         return (toJSON (error));
      }
      else
      {
         Double psd = statsSvc.PopulationStandardDeviation(values);
         CalculatorAnswer answer = new CalculatorAnswer(true, "Population Standard Deviation", psd);
         return (toJSON (answer));
      }
   }

   // ----------------------------------------------------------------------
   //
   @PostMapping(path="/CalculateMean", consumes = MediaType.APPLICATION_JSON_VALUE,
           produces = MediaType.APPLICATION_JSON_VALUE)
   public String CalculateMean(@RequestBody String data)
   {
      List<Double> values = statsSvc.IsDataValidForMean(data);
      if (values == null)
      {
         CalculatorError error = new CalculatorError(false, "Invalid Input", "Mean format is one number per line");
         return (toJSON (error));
      }
      else
      {
         Double mean = statsSvc.Mean(values);

         CalculatorAnswer answer = new CalculatorAnswer(true, "Mean", mean);
         return (toJSON (answer));
      }
   }

   // ----------------------------------------------------------------------
   //
   @PostMapping(path="/CalculateZScore", consumes = MediaType.APPLICATION_JSON_VALUE,
           produces = MediaType.APPLICATION_JSON_VALUE)
   public String CalculateZScore(@RequestBody String data)
   {
      ZScoreInput z = statsSvc.IsDataValidForZScore(data);

      if (z == null)
      {
         CalculatorError error = new CalculatorError(false, "Invalid Input", "Z-Score format is \"value, mean, stdDev\" on one line separated by commas");
         return (toJSON (error));
      }
      else
      {
         Double zScore = statsSvc.ZScore(z);
         CalculatorAnswer answer = new CalculatorAnswer(true, "Z-Score", zScore);
         return (toJSON (answer));
      }
   }

   // ----------------------------------------------------------------------
   //
   @PostMapping(path="/CalculateLinearRegression", consumes = MediaType.APPLICATION_JSON_VALUE,
           produces = MediaType.APPLICATION_JSON_VALUE)
   public CalculatorResponse CalculateLinearRegression(@RequestBody String data)
   {
      List<XYPair> values = linearRegSvc.IsDataValidForLinearRegression(data);
      if (values == null)
         return (new CalculatorError(false, "Invalid Input", "Simple Linear Regression format is one x, y value per line"));
      else
      {
         String formula = linearRegSvc.CalculateLinearRegression(values);
         return (new CalculatorAnswer(true, "Single Linear Regression Formula", formula));
      }
   }

   // ----------------------------------------------------------------------
   //
   @PostMapping(path="/CalculateYValue", consumes = MediaType.APPLICATION_JSON_VALUE,
           produces = MediaType.APPLICATION_JSON_VALUE)
   public CalculatorResponse CalculateYValue(@RequestBody String data)
   {
      XMBInput input = linearRegSvc.IsDataValidForYValue(data);
      if (input == null)
         return (new CalculatorError(false, "Invalid Input", "Predict Y format is \"x,m,b\" on one line separated by commas"));
      else
      {
         Double yValue = linearRegSvc.CalculateYValue(input);
         return (new CalculatorAnswer(true, "Single Linear Regression Prediction", yValue));
      }
   }
}
